File: Analysis/Ex/LoKiExample/src/LoKi_Track2MC.cpp
// $Id: LoKi_Track2MC.cpp,v 1.1 2010-05-06 11:14:48 ibelyaev Exp $
// ============================================================================
// Include files 
// ============================================================================
// LoKi 
// ============================================================================
#include "LoKi/Algo.h"
// ============================================================================
// Event 
// ============================================================================
#include "Event/MCParticle.h"
// ============================================================================
// Relations
// ============================================================================
#include "Relations/IRelationWeighted2D.h"
// ============================================================================
// DaVinciMCKernel
// ============================================================================
#include "Kernel/Track2MC.h"
// ============================================================================
/** @file 
 *  Simple algorithm to illustrate the usage of  Track<-->MC relation tables 
 *
 *  This file is a part of LoKi project - 
 *    "C++ ToolKit  for Smart and Friendly Physics Analysis"
 *
 *  The package has been designed with the kind help from
 *  Galina PAKHLOVA and Sergey BARSUK.  Many bright ideas, 
 *  contributions and advices from G.Raven, J.van Tilburg, 
 *  A.Golutvin, P.Koppenburg have been used in the design.
 *
 *  @author Vanya BELYAEV Ivan.Belyaev@itep.ru
 *  @date 2010-05-06
 */
// ============================================================================
LOKI_ALGORITHM(Track2MCEx)
{
  
  // get Tracks from Transient Event Store  
  LHCb::Track::Range tracks = 
    get<LHCb::Track::Range> ( LHCb::TrackLocation::Default    ) ;
  
  // get the relation table form Transient Event Store  
  const LHCb::Track2MC2D* table2d  = 
    get <LHCb::Track2MC2D>  ( LHCb::Track2MCLocation::Default ) ;
  
  // get "inverse" table : MCParticle -> Track 
  const LHCb::MC2Track* itable = table2d->inverse() ;
  
  // Track -> MCParticle links 
  
  // loop over the tracks:
  for ( LHCb::Track::Range::iterator itrack = tracks.begin() ; 
        tracks.end() != itrack ; ++itrack ) 
  {
    const LHCb::Track* track = *itrack ;
    if ( 0 == track    ) { continue ; }                          // CONTINUE  
    
    // get all relations for the given track:
    LHCb::Track2MC2D::Range links = table2d->relations ( track ) ;
    if ( links.empty() ) { continue ; }                          // CONTINUE 
    
    // loop over related MC-particles :
    for ( LHCb::Track2MC2D::Range::iterator link = links.begin() ; 
          links.end() != link ; ++link ) 
    {
      // get the related MC-particle 
      const LHCb::MCParticle* mcparticle = link->to() ;
      // get the "weight" associated with the given link:
      const double weight = link->weight() ;
      // print: 
      // ...
      info() << "The key/weight for related MC-particle " 
             << mcparticle->key() << "/"  
             << track->key()      << "/" << weight << endmsg ;
      
    } //                                            end of the loop over links

  } //                                              end of the loop over tracks
  
  
  // MCParticle -> Track links
  
  //  get all MC-particles from Transient store 
  LHCb::MCParticle::Container* mcparticles = 
    get<LHCb::MCParticle::Container> ( LHCb::MCParticleLocation::Default ) ;
  
  // loop over all MC-particles:
  for ( LHCb::MCParticle::Container::iterator imc = mcparticles->begin() ; 
        mcparticles->end() != imc ; ++imc ) 
  {
    const LHCb::MCParticle* mc = *imc ;
    if ( 0 == mc ) { continue ; }                                // CONTINUE 
    
    // get all related tracks:
    LHCb::MC2Track::Range links = itable->relations ( mc ) ;
    if ( links.empty() ) { continue ; }                          // CONTINUE 
    
    // links over all related tracks:
    for (  LHCb::MC2Track::Range::iterator link = links.begin() ; 
           links.end() != link ; ++link ) 
    {
      // get the related track
      const LHCb::Track* track  = link->to() ;
      // get the "weight" associated with the given link:
      const double       weight = link->weight() ;
      // print
      // ...
      info() << "The key/weight for related Track " 
             << mc    -> key () << "/"  
             << track -> key () << "/" << weight << endmsg ;
      
    } //                                 end of the loop over links

  } //                                   end of the loop over all MC-particles 
  
  
  return StatusCode::SUCCESS ;
  
}
// ============================================================================
// The END 
// ============================================================================

File: Analysis/Ex/LoKiExample/python/LoKiExample/LoKi_Track2MC.py
#!/usr/bin/env gaudirun.py
# =============================================================================
# $Id: LoKi_Track2MC.py,v 1.1 2010-05-06 11:14:48 ibelyaev Exp $ 
# =============================================================================
## @file
#  The configuration file to run LoKi_Track2MC example
#
#   This file is a part of LoKi project - 
#     "C++ ToolKit  for Smart and Friendly Physics Analysis"
#
#  The package has been designed with the kind help from
#  Galina PAKHLOVA and Sergey BARSUK.  Many bright ideas, 
#  contributions and advices from G.Raven, J.van Tilburg, 
#  A.Golutvin, P.Koppenburg have been used in the design.
#
#  By usage of this code one clearly states the disagreement 
#  with the campain of Dr.O.Callot et al.: 
#  "No Vanya's lines are allowed in LHCb/Gaudi software."
#
#  @author Vanya BELYAEV Ivan.Belyaev@itep.ru
#  @date   2010-05-06
# =============================================================================
"""
Configuration file for LoKiExample package

This file is a part of LoKi project - 
\"C++ ToolKit  for Smart and Friendly Physics Analysis\"

The package has been designed with the kind help from
Galina PAKHLOVA and Sergey BARSUK.  Many bright ideas, 
contributions and advices from G.Raven, J.van Tilburg, 
A.Golutvin, P.Koppenburg have been used in the design.

By usage of this code one clearly states the disagreement 
with the campain of Dr.O.Callot et al.: 
\"No Vanya's lines are allowed in LHCb/Gaudi software.\"

"""
# =============================================================================
__author__  = " Vanya BELYAEV Ivan.Belyaev@itep.ru "
__version__ = " CVS Tag $Name: not supported by cvs2svn $, version $Revision: 1.1 $ "
__date__    = " 2010-05-06 "
# =============================================================================

from  Gaudi.Configuration import * 

## create & configure own algorithm:
from  Configurables import LoKi__Track2MCEx
alg = LoKi__Track2MCEx (
    "Track2MCEx"
    )

## configure Track<-->MC relation tables
import LoKiPhysMC.Track2MC_Configuration

## get input data:
from LoKiExample.Bs2Jpsiphi_mm_data import Inputs as INPUT 

## confgure the application itself:
from  Configurables import DaVinci 
DaVinci (
    DataType       = 'DC06'  , ## Data type  
    Simulation     = True    , ## Monte Carlo 
    Hlt            = False      ,
    #
    UserAlgorithms = [ alg ] , ## let DaVinci know about local algorithm
    # delegate this properties to Event Selector 
    EvtMax        = 20       ,  
    SkipEvents    = 0        ,
    Input         = INPUT    , ## the list of input data files
    # delegate to Histogram Persistency Service
    HistogramFile = "Phi_Histos.root"
    )

# =============================================================================
# The END
# =============================================================================
